### PR TITLE
fix: improve health checks

### PR DESCRIPTION
**Pull request by tzneal**
**Description**
Acquire and release some locks during health checks to identify deadlock.
**How was this change tested?**

*

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [X] No

**Release Note**

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: pkg/controllers/state/cluster.go [hunk 1]
```diff
@@ -512,3 +513,10 @@ func (c *Cluster) Synchronized(ctx context.Context) error {
 func (c *Cluster) recordConsolidationChange() {
 	atomic.StoreInt64(&c.consolidationState, c.clock.Now().UnixMilli())
 }
+
+func (c *Cluster) LivenessProbe(req *http.Request) error {
+	c.mu.RLock()
```
Comment (bwagner5) [inline pkg/controllers/state/cluster.go:518]: why not take a write lock here?

**Review by bwagner5 — APPROVED** [review]
nicee!

File: pkg/controllers/controllers.go [hunk 2]
```diff
@@ -113,9 +113,11 @@ func Initialize(injectCloudProvider func(context.Context, cloudprovider.Options)
 		utilruntime.Must(registerPprof(manager))
 	}
 
-	cloudProvider := cloudprovidermetrics.Decorate(
-		injectCloudProvider(ctx, cloudprovider.Options{ClientSet: clientSet, KubeClient: manager.GetClient(), StartAsync: manager.Elected()}),
-	)
+	cloudProvider := injectCloudProvider(ctx, cloudprovider.Options{ClientSet: clientSet, KubeClient: manager.GetClient(), StartAsync: manager.Elected()})
+	if hp, ok := cloudProvider.(HealthCheck); ok {
```
Comment (ellistarn) [inline pkg/controllers/controllers.go:117]: Another option is to write an AWS specific controller that checks the liveness of the cloudprovider. I've been wanting to build a simple mechanism to wire additional controllers through Initialize()

File: pkg/controllers/state/cluster.go (hunk 1 above)
Comment (tzneal) [reply pkg/controllers/state/cluster.go:518]: Yeah, that's better. I did the read locks at first to be less intrusive, but I agree that write locks is better at finding the problem quickly.

File: pkg/controllers/controllers.go (hunk 2 above)
Comment (tzneal) [reply pkg/controllers/controllers.go:117]: I was hoping to make these health checks better when I started this, but couldn't really come up with much apart from preventing deadlock.  If we can come up with any scenario where restarting is the best corrective action, we can just hook it up to a health check and return it as an error which will provide a bit more info on why we restarted.

**Review by ellistarn — DISMISSED** [review]
Let's hold for the release. But great change!

**Review by bwagner5 — APPROVED** [review]
lgtm!

**Pull request merged**